New Android Custom Tabs Guide (#6122)
GoogleChrome/developer.chrome.com: site/en/docs/android/custom-tabs/guide-warmup-prefetch/index.md

---
layout: "layouts/doc-post.njk"

> title: "Warm-up and pre-fetching: using the Custom Tabs Service"

Reviewer: I thought we had an easier way to do this with `android-browser-helper`, but can't find it (I thought maybe in the `TwaLauncher`). Long-term, an idea would to add that to that library.

Author: There's an easier way, but not enough to justify the overhead of explaining why to add another library.

> title: "Warm-up and pre-fetching: using the Custom Tabs Service"

Reviewer 2: The items before the colon need to be the same part of speech: either "Warming-up and pre-fetching" or "Warm-up and pre-fetch".

Author: fixed

seoTitle: "Guide: warm-up and pre-fetching - using the Custom Tabs Service"
date: 2023-04-21
description: Learn how to make sure that your Custom Tabs start as quickly as possible.
authors:
  - sebastianbenz
---

> The third part of this guide focuses on warming up the browser process via [warmup()](https://developer.android.com/reference/androidx/browser/customtabs/CustomTabsClient#warmup(long)) and prefetching web pages via [mayLaunchUrl](https://developer.android.com/reference/androidx/browser/customtabs/CustomTabsSession#mayLaunchUrl(android.net.Uri,android.os.Bundle,java.util.List%3Candroid.os.Bundle%3E))(). Warming up the browser process can save up to 700ms when opening a link. Pre-rendering content via mayLaunchUrl will make external content open instantly. Together both APIs can greatly improve the user experience of a Custom Tabs integration and are highly recommended.

Reviewer 3: > The third part of this guide focuses on warming up the browser process via [warmup()]

nit: warming up via wamup is meaningless if you wont know what warming up a browser means. Would it make sense to add a quick summary - something like

> The third part of this guide focuses on speeding up the browser startup via [warmup()]

Author: good suggestion! done

> The third part of this guide focuses on warming up the browser process via [warmup()](https://developer.android.com/reference/androidx/browser/customtabs/CustomTabsClient#warmup(long)) and prefetching web pages via [mayLaunchUrl](https://developer.android.com/reference/androidx/browser/customtabs/CustomTabsSession#mayLaunchUrl(android.net.Uri,android.os.Bundle,java.util.List%3Candroid.os.Bundle%3E))(). Warming up the browser process can save up to 700ms when opening a link. Pre-rendering content via mayLaunchUrl will make external content open instantly. Together both APIs can greatly improve the user experience of a Custom Tabs integration and are highly recommended.

Reviewer 2: ```suggestion
The third part of this guide focuses on warming up the browser process via [`warmup()`](https://developer.android.com/reference/androidx/browser/customtabs/CustomTabsClient#warmup(long)) and prefetching web pages via [`mayLaunchUrl()`](https://developer.android.com/reference/androidx/browser/customtabs/CustomTabsSession#mayLaunchUrl(android.net.Uri,android.os.Bundle,java.util.List%3Candroid.os.Bundle%3E)). Warming up the browser process can save up to 700ms when opening a link. Pre-rendering content via `mayLaunchUrl()` will make external content open instantly. Together both APIs can greatly improve the user experience of a Custom Tabs integration and are highly recommended.
```

> The basic steps are:

Reviewer 3: nit: google styleguides say to avoid phrases like "just" "basic" and "simple"

> 1. Check via [CustomTabsClient](https://developer.android.com/reference/androidx/browser/customtabs/CustomTabsClient).[getPackageName()](https://developer.android.com/reference/androidx/browser/customtabs/CustomTabsClient#getPackageName(android.content.Context,java.util.List%3Cjava.lang.String%3E)) If the default browser supports Custom Tabs. If yes, bind to the CustomTabsService via [CustomTabsClient.bindCustomTabsService()](https://developer.android.com/reference/androidx/browser/customtabs/CustomTabsClient#bindCustomTabsService(android.content.Context,java.lang.String,androidx.browser.customtabs.CustomTabsServiceConnection)).

Reviewer 2: ```suggestion
1. Check via [`CustomTabsClient`](https://developer.android.com/reference/androidx/browser/customtabs/CustomTabsClient).[`getPackageName()`](https://developer.android.com/reference/androidx/browser/customtabs/CustomTabsClient#getPackageName(android.content.Context,java.util.List%3Cjava.lang.String%3E)) If the default browser supports Custom Tabs. If yes, bind to the CustomTabsService via [`CustomTabsClient.bindCustomTabsService()`](https://developer.android.com/reference/androidx/browser/customtabs/CustomTabsClient#bindCustomTabsService(android.content.Context,java.lang.String,androidx.browser.customtabs.CustomTabsServiceConnection)).
```

Reviewer 2: >If the default browser supports Custom Tabs.

This isn't a complete sentence. Is this supposed to be part of the previous sentence or is something else missing?

Author: part of the previous sentence, fixed

> 2. Once connected to the CustomTabsService, in the [CustomTabsServiceConnection.onCustomTabsServiceConnected()](https://developer.android.com/reference/androidx/browser/customtabs/CustomTabsServiceConnection#onCustomTabsServiceConnected(android.content.ComponentName,androidx.browser.customtabs.CustomTabsClient)) callback, do:

Reviewer 2: ```suggestion
2. Once connected to the CustomTabsService, in the [`CustomTabsServiceConnection.onCustomTabsServiceConnected()`](https://developer.android.com/reference/androidx/browser/customtabs/CustomTabsServiceConnection#onCustomTabsServiceConnected(android.content.ComponentName,androidx.browser.customtabs.CustomTabsClient)) callback, do:
```

>     1. Warmup the browser process via [CustomTabsClient.warmup()](https://developer.android.com/reference/androidx/browser/customtabs/CustomTabsClient#warmup(long)).

Reviewer 2: ```suggestion
    a. Warmup the browser process via [`CustomTabsClient.warmup()`](https://developer.android.com/reference/androidx/browser/customtabs/CustomTabsClient#warmup(long)).
```

>     2. Create a new [CustomTabsSession](https://developer.android.com/reference/androidx/browser/customtabs/CustomTabsSession) via [CustomTabsClient.newSession()](https://developer.android.com/reference/androidx/browser/customtabs/CustomTabsClient#newSession(androidx.browser.customtabs.CustomTabsCallback,int)).

Reviewer 2: ```suggestion
    a. Create a new [`CustomTabsSession`](https://developer.android.com/reference/androidx/browser/customtabs/CustomTabsSession) via [`CustomTabsClient.newSession()`](https://developer.android.com/reference/androidx/browser/customtabs/CustomTabsClient#newSession(androidx.browser.customtabs.CustomTabsCallback,int)).
```

> 3. Optionally, prefetch web pages the user is likely to visit via [CustomTabsSession.mayLaunchUrl](https://developer.android.com/reference/androidx/browser/customtabs/CustomTabsSession#mayLaunchUrl(android.net.Uri,android.os.Bundle,java.util.List%3Candroid.os.Bundle%3E))().

Reviewer 2: ```suggestion
3. Optionally, prefetch web pages the user is likely to visit via [`CustomTabsSession.mayLaunchUrl()`](https://developer.android.com/reference/androidx/browser/customtabs/CustomTabsSession#mayLaunchUrl(android.net.Uri,android.os.Bundle,java.util.List%3Candroid.os.Bundle%3E)).
```

> 4. When launching a new Custom Tab, pass the [CustomTabsSession](https://developer.android.com/reference/androidx/browser/customtabs/CustomTabsSession) to the [CustomTabsIntent.Builder](https://developer.android.com/reference/androidx/browser/customtabs/CustomTabsIntent.Builder) via the constructor `new CustomTabsIntent.Builder(session)`.

Reviewer 3: Personally, these steps as comments throughout the code rather than a separate section would be more helpful

Author: yeah, I added this because you can have links

> 4. When launching a new Custom Tab, pass the [CustomTabsSession](https://developer.android.com/reference/androidx/browser/customtabs/CustomTabsSession) to the [CustomTabsIntent.Builder](https://developer.android.com/reference/androidx/browser/customtabs/CustomTabsIntent.Builder) via the constructor `new CustomTabsIntent.Builder(session)`.

Reviewer 2: ```suggestion
4. When launching a new Custom Tab, pass the [`CustomTabsSession`](https://developer.android.com/reference/androidx/browser/customtabs/CustomTabsSession) to the [CustomTabsIntent.Builder](https://developer.android.com/reference/androidx/browser/customtabs/CustomTabsIntent.Builder) via the constructor `new CustomTabsIntent.Builder(session)`.
```

> Important: when targeting API level 30, [CustomTabsClient](https://developer.android.com/reference/androidx/browser/customtabs/CustomTabsClient).[getPackageName()](https://developer.android.com/reference/androidx/browser/customtabs/CustomTabsClient#getPackageName(android.content.Context,java.util.List%3Cjava.lang.String%3E)) requires you to add a queries section to your Android Manifest, declaring an intent-filter that matches browsers with Custom Tabs support.

Reviewer 2: ```suggestion
{% Aside 'important' %}
When targeting API level 30, [`CustomTabsClient`](https://developer.android.com/reference/androidx/browser/customtabs/CustomTabsClient).[`getPackageName()`](https://developer.android.com/reference/androidx/browser/customtabs/CustomTabsClient#getPackageName(android.content.Context,java.util.List%3Cjava.lang.String%3E)) requires you to add a queries section to your Android Manifest, declaring an intent-filter that matches browsers with Custom Tabs support.
{% endAside %}
```

```xml
<manifest xmlns:android="http://schemas.android.com/apk/res/android"
    xmlns:tools="http://schemas.android.com/tools">
  …
   <queries>
        <intent>
            <action android:name="android.support.customtabs.action.CustomTabsService" />
        </intent>
    </queries>
</manifest>
```

> Here is the full code:

Reviewer 3: nit: this sentence is not needed. i'd remove it.

Author: I like to have something that separates the two snippets. rephrased

```java
private CustomTabsClient mClient;
private CustomTabsSession mSession;

> private CustomTabsServiceConnection mConnection= new CustomTabsServiceConnection() {

Author: done

    @Override
    public void onCustomTabsServiceConnected(
            @NonNull ComponentName name,
            @NonNull CustomTabsClient client
    ) {
        mClient = client;
        // Warm up the browser process
        mClient.warmup(0 /* placeholder for future use */);
        // Create a new browser session
        mSession = mClient.newSession(new CustomTabsCallback());

>         // Optional: pre-render pages the user is likely to visit

Reviewer 3: all of this is optional, no?

Author: removed

        // Do nothing if there is an existing service connection
        return;
    }

    // Get the default browser package name, this will be null if
    // the default browser does not provide a CustomTabsService
    String packageName = CustomTabsClient.getPackageName(context, null);
    if (packageName == null) {
        // Do nothing as service connection is not supported
        return;
    }
    CustomTabsClient.bindCustomTabsService(context, packageName, mConnection);
}

@Override
protected void onCreate(Bundle savedInstanceState) {
…
    bindCustomTabService(this);
    findViewById(R.id.button).setOnClickListener(new View.OnClickListener() {
        @Override
        public void onClick(View view) {
            String url = "https://developers.android.com";
            CustomTabsIntent intent = new CustomTabsIntent.Builder(session)
                    .build();
            intent.launchUrl(MainActivity.this, Uri.parse(url));
        }
    });
}
```

{% Aside "caution" %}
A Custom Tabs service connection might fail while your activity is running. If your app requires an active service connection, there are two strategies to ensure a working service connection:

> 1. When eagerly connecting to the CustomTabsService during `onCreate`, reconnect to the service  if your activity gets disconnected and the `onServiceDisconnected` callback gets invoked.

Reviewer 2: ```suggestion
1. When eagerly connecting to the `CustomTabsService` during `onCreate()`, reconnect to the service  if your activity gets disconnected and the `onServiceDisconnected()` callback gets invoked.
```

Reviewer 2: >When eagerly connecting

What are you trying to convey with this?

Author: Connecting to the service before the user requests a custom tab - I think this should be clear to an android developer as I also mention oncreate.

Reviewer 2: It's the word 'eagerly' I'm having trouble with. Technology can't be eager.

Author: It's a common programming idiom usually used in the context of evaluation https://en.wikipedia.org/wiki/Evaluation_strategy#Eager_evaluation

> 2. Wait until the user hits a button, establish the service connection and then launch the custom tab.

Reviewer 3: > Wait until the user hits a button

do you mean triggers an action to launch a CustomTab? I am not sure I understand what button is being hit

Author: rephrased
